### PR TITLE
Replace hoedown with pull in rustdoc

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
cc @rust-lang/docs

### CODE REVIEW

File: src/librustdoc/html/markdown.rs
Comment (steveklabnik) [inline src/librustdoc/html/markdown.rs]: Why leave this in but commented out?

File: src/librustdoc/html/markdown.rs
Comment (steveklabnik) [inline src/librustdoc/html/markdown.rs]: same here

**Review by steveklabnik — COMMENTED** [review]
I will give this a real review tomorrow, but 🎊

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: If you made this:

```rust
let event = match next_event {
    Some(event) => event,
    None => break,
};
....
```

This would reduce the indentation

Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: At that point, couldn't this be a:

```rust
while let Some(event) = parser.next() {
    ...
}
```

**Review by steveklabnik — COMMENTED** [review]
Seems good to me, though I agree with @frewsxcv 's comments about nesting

**Comment by GuillaumeGomez** [comment]
This is now ready to review!

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: Could this could just be a `for` loop?

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: `for` loop?

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: `for` loop?

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: Pull in rust-url for this?

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: Could return early here if this is None

File: src/librustdoc/html/markdown.rs
Comment (steveklabnik) [inline src/librustdoc/html/markdown.rs]: yeah, boolean parameters are usually a code smell

File: src/Cargo.lock
Comment (killercup) [inline src/Cargo.lock]: 0.0.8… Another lib we should help get to 1.0

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: Is this the only place where `shorter` is actually used? The name `shorter` does not imply "no tbody tag" to me…

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: I think I'd prefer an `enum MarkdownOutputStyle { Compact, Fancy }` (or some other fitting variant names) instead of `bool`

**Comment by steveklabnik** [comment]
So, I'm spot-checking some text to find regressions.

std::collections module page

collections


std::convert module page

convert

prelude:

prelude

some smooshed text

std::sync::LockResult

lock

std::string

string

that's what i found just now. we may need to tweak some text, and it's probably good to go through the whole entire thing once the technical side is good and clear up these kinds of issues.

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: I'd nest the `if let` and the `match`. It's hard to see that only the second match arm is actually interesting

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: Agreed, this is `for event in parser { match event { … } }`

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: finally, a good reason not use a for loop ;)

**Review by killercup — COMMENTED** [review]
Nice to see some progress here!

One pattern I immediately saw was all the `loop` constructs matching the event and tag. I'd:

1. `use Event::*` and `use Tag::*;` to reduce the visual noise
2. Try to rewrite them as `let events = parser.$filter_relevant_events; for event in events { match event {…} }`, where `$filter_relevant_events` are some operations on the iterator to get the events you are interested in (e.g., to use `take_while` instead of adding `break` arms).

    FYI, I wrote a macro once for this exact use case to allow `events.take_while(not!(end Tag::Paragraph))` and similar. This may help converting most of these loops into nice iterators.

    You can find it and example using it [here](https://github.com/killercup/rust-docstrings/blob/ae50902a298270c929b8ca8325537e6cfca2f041/src/extractors.rs#L8-L30). You'll probably need to extend it to support multiple end tag variants as used below.

I'd also really try to get rid of the `shorter: bool` stuff and replace with some semantically meaningful names :)

Switching parser is not something I expect to just work, even when both apparently conform to the CommonMark spec. Before merging this, I'd render a bunch of docs (std and some larger crates), and try to

- set up a script that opens and takes screenshots of all generated html pages to visually diff them with ones generated by the current stable rustdoc,
- or let humans look at the rendered pages,
- or do both 😄

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: This is not part of the change, however I can do it next up.

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: @killercup: Good idea!

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: In paragraph as well. The point is to display only the first line, so everything that could be on more than one line use it.

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: Doesn't change much from my point of view... I think this syntax is actually better.

**Comment by GuillaumeGomez** [comment]
Most problems should be solved now. I'll write a macro next to allow to remove all these loops.

**Comment by ollie27** [comment]
Is there any reason you're not using `pulldown_cmark::html::push_html` to actually render the HTML?

**Comment by GuillaumeGomez** [comment]
@ollie27: Unfortunately yes: the generation of urls mainly.

File: src/librustdoc/html/markdown.rs
Comment (steveklabnik) [inline src/librustdoc/html/markdown.rs]: shouldn't this mean the comment above goes away too?

Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: Absolutely!

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: Nice. I'd call it `collect_md_events`, though, and make calling it a bit more expressive, e.g.

```rust
collect_md_events!(from parser into &mut buf (toc_builder, shorter) until End(Header(_)))
```

File: src/librustdoc/html/markdown.rs
Comment (killercup) [inline src/librustdoc/html/markdown.rs]: Could also be using the macro (e.g. when calling without toc builder and shorter)

**Review by killercup — COMMENTED** [review]
Code looks nicer with the macro, though I'd probably make it more generic.

Also, why not `use Event::*` and `use Tag::*`?

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: Not sure if this is super useful. Anyone else has an opinion on it?

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: I didn't see much interest in expanding the macro to handle this case. But yes it's possible.

**Comment by GuillaumeGomez** [comment]
> Also, why not use `Event::*` and use `Tag::*`?

Simply because I like the code being explicit. At least with this syntax, you don't have to think about it, the information is already there.

**Comment by ollie27** [comment]
> @ollie27: Unfortunately yes: the generation of urls mainly (for play.rust-lang in blockcodes).

You can still use `push_html` and add these links by implementing an iterator adapter around `Parser`. It would pass most `Event`s though but you could filter and modify the ones you're interested in like `CodeBlock` and `Header`. This way you don't have to reimplement all the HTML rendering that `pulldown-cmark` already provides.

**Comment by GuillaumeGomez** [comment]
@ollie27: That's what I thought at first but then the "compact markdown" thing appeared.

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: This can be a `for` loop, right?

File: src/librustdoc/html/markdown.rs
Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: `for` loop?

Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: If so, there are many other places in this PR `for` loops can be used

Comment (frewsxcv) [inline src/librustdoc/html/markdown.rs]: Maybe not actually because it gets passed in `looper`?

**Comment by steveklabnik** [comment]
@GuillaumeGomez it doesn't seem normal to me, but i re-did it and it failed the same way again.

**Comment by GuillaumeGomez** [comment]
Strange... The worst part is that we don't have any information.

**Comment by QuietMisdreavus** [comment]
cc #38400

**Comment by GuillaumeGomez** [comment]
cc @rust-lang/core

**Comment by GuillaumeGomez** [comment]
cc @alexcrichton (I think you might be the one which could help the most in here.)

**Comment by alexcrichton** [comment]
What am I being cc'd on here for?

**Comment by GuillaumeGomez** [comment]
@alexcrichton: Because you worked on the buildchain and you're in the tools team. Apparently there is an in run-make which might be a side effect of this changes. However, I don't understand what's the link with this code change. If you have no clue, do you know who might be able to help on this please?

**Comment by alexcrichton** [comment]
Did you cc the core team b/c your PR is failing tests? 

There's also no links to any failures? Are you wondering about the current travis failure? In general we're all quite busy so cc'ing us and hoping we can figure out why is generally a recipe for a failed cc ...

**Comment by GuillaumeGomez** [comment]
I was asking for opinions because I wasn't able to find anything. So maybe it rings a bell to someone or someone who wrote the run-make tests might have an idea why a change in rustdoc breaks it since it doesn't seem to be called in there. If you have no clue but might know someone who might, I'd ask him directly.

**Comment by steveklabnik** [comment]
Okay, now we're back to the strange failure: https://travis-ci.org/rust-lang/rust/jobs/213447223#L3177

```text
make[1]: Entering directory '/checkout/src/test/run-make/sysroot-crates-are-unstable'

verifying getopts is an unstable crate

crate getopts is not unstable

error[E0464]: multiple matching crates for `getopts`

 --> <anon>:1:1

  |

1 | extern crate getopts;

  | ^^^^^^^^^^^^^^^^^^^^^

  |

  = note: candidates:

  = note: path: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-a80f7c589731de53.so

  = note: path: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-a80f7c589731de53.rlib

  = note: crate name: getopts

  = note: path: /checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib/libgetopts-ccd85c85236add46.rlib

  = note: crate name: getopts
```

I don't see this happen locally, but I am on Windows, not Linux where the build is.

**Comment by alexcrichton** [comment]
Ah yes so that bug has to do with pulling in dependencies through rustbuild. Because compiler/rustdoc dependencies work differently than normal dependencies (they go in and are linked through the sysroot) they have to be treated a little specially. The bug here is that pulldown-cmark (the crate) depends on getopts from crates.io. This means that we're compiling two copies of getopts, one from crates.io and one in the compiler itself. There's currently no great way to handle this other than removing the getopts dependency from pulldown-cmark.

**Comment by alexcrichton** [comment]
Note that there are other restrictions of pulling in crates.io crates into the build. I'd really highly recommend running the test suite locally as it will tell you everything you need to know, but more info can be found in the [`log` crate removal](https://github.com/rust-lang/rust/pull/40347)

**Comment by steveklabnik** [comment]
> I'd really highly recommend running the test suite locally as it will tell you everything you need to know, 

I thought I was, I probably got the `x.py` incantation incorrect, bah.

**Comment by GuillaumeGomez** [comment]
Same for me... Any way, I'm currently trying to remove the `getops` dependency when unneeded (which seems problematic because of cargo bug for the `--no-default-features` option).

File: src/librustdoc/html/markdown.rs
Comment (steveklabnik) [inline src/librustdoc/html/markdown.rs]: this all feels like acceptable "breakage" to me, nobody is gonna be relying on single vs double quote headings here, nor should they

**Review by steveklabnik — COMMENTED** [review]
I think this looks good to me. @frewsxcv ?

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: I actually had issue because `rustc` tried to interpret `\'` as an unicode character. Like this, this issue is definitely solved.

File: src/librustdoc/html/markdown.rs
Comment (ScottAbbey) [inline src/librustdoc/html/markdown.rs]: I can't tell whether or not the documentation for this is rendered anywhere (?), but it seems like this comment should be updated, regardless.  It talks about the "hoedown C-library" and "the C bindings."  Also, the example might not work anymore, it uses `Markdown(s)`, but that appears to now be `Markdown(s, style)`

File: src/librustdoc/html/markdown.rs
Comment (ScottAbbey) [inline src/librustdoc/html/markdown.rs]: Consider calling this `code` instead?

File: src/librustdoc/html/markdown.rs
Comment (ScottAbbey) [inline src/librustdoc/html/markdown.rs]: Consider calling this function `code_block` or similar.  I found this name confusing because "block" has a totally different meaning within pulldown-cmark and the CommonMark specification.  There, "block" is a general term that refers to paragraphs, block quotations, lists, headings, and code blocks, they're all called "blocks".

**Review by ScottAbbey — COMMENTED** [review]
These are all inner functions but I feel like naming is important everywhere.

**Comment by GuillaumeGomez** [comment]
@ScottAbbey: Your remarks make sense. I'll apply your suggestions then!

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: Excellent point!

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: I'll go for `inline_code`.

File: src/librustdoc/html/markdown.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/markdown.rs]: I prefer to be explicit so I went for `table_`.

**Comment by GuillaumeGomez** [comment]
Updated.

**Review by frewsxcv — APPROVED** [review]
I have a bunch of nitpicky style things, but they aren't important enough to hold up this PR any longer. Looks good to me!

**Comment by GuillaumeGomez** [comment]
@frewsxcv, @steveklabnik: Since both of you agreed on the changes, I'll r+ this PR. Don't hesitate to r- if you find anything!

@bors: r=frewsxcv,steveklabnik

**Comment by ollie27** [comment]
Was this actually finished? It's missing support for horizontal rules, footnotes and images. There are also several bugs like missing the `<br />` for `HardBreak`s, the title part of links being rendered as visible text rather than as an attribute on the `<a>` tag and `MarkdownHtml` now doing the same as `Markdown` when it should be escaping raw HTML.

As I previously mentioned, using `push_html` would have been much easier than trying to reimplement all of the HTML rendering.

**Comment by frewsxcv** [comment]
> Was this actually finished? It's missing support for horizontal rules, footnotes and images. There are also several bugs like missing the `<br />` for `HardBreak`s, the title part of links being rendered as visible text rather than as an attribute on the `<a>` tag and `MarkdownHtml` now doing the same as `Markdown` when it should be escaping raw HTML.

Have you confirmed any of these? I can bring up these concerns during today's Documentation Team meeting.

**Comment by ollie27** [comment]
> Have you confirmed any of these?

Yes. Try rendering the following in the current rustdoc and after this change:

```rust
/// markdown test
///
/// this is a [link].
///
/// [link]: https://example.com "this is a title"
///
/// hard break:  
/// after hard break
///
/// -----------
///
/// a footnote[^footnote].
///
/// [^footnote]: Thing
///
/// ![Rust](https://www.rust-lang.org/logos/rust-logo-128x128-blk-v2.png)
#[deprecated(note = "Struct<T>")]
pub fn f() {}
```

Before:
Before

After:
After

**Comment by steveklabnik** [comment]
Two things:

1. this landed after last night's nightly, so tonight's is the first night where it lands. I was already planning on trying to publicize "please look at this since it's a big change", so knowing this stuff is good.
2. I am currently building docs with the commit before and the commit after, and then i'm going to diff them to see exactly what's different.

I don't think that this missing a few things is the end of the world; now that this has landed, fixing stuff up is much easier. We've still got plenty of time before the release; four weeks until beta, and then we could backport anything catastrophic. Getting this out in the wild sooner rather than later is important IMHO.

**Comment by steveklabnik** [comment]
Here is an un-semantic diff https://github.com/steveklabnik/docdiff/commit/5d1706170c2aeb1bc61d36cd84a27d8ed5880cf2

**Comment by steveklabnik** [comment]
A much prettier diff https://github.com/steveklabnik/docdiff/compare/HEAD~1...master

**Comment by steveklabnik** [comment]
I've filed https://github.com/rust-lang/rust/issues/40912 to track these regressions. 😄

**Comment by mbrubeck** [comment]
This was disabled by default in #41431, and #44229 is tracking the plan to enable it by default.

**Pull request merged**
